render_bundle/lib/charm_bundle_generator.py

```diff
@@ -0,0 +1,84 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Generates any charm bundle from user provided Application and Relation objects."""
+
+from pathlib import Path
+from typing import Optional
+
+from jinja2 import Template
+from pydantic import BaseModel
+
+
+def write_to_file(file: str, content: str):
+    """Writes content to a file.
+
+    Args:
+        file: File name
+        content: File content
+    """
+    with open(file, "w") as f:
+        f.write(content)
+
+
+class Resource(BaseModel):
+    """Pydantic model describing a charm application resource."""
+
+    name: str
+    value: str
+
+
+class Application(BaseModel):
+    """Pydantic model describing a charm application."""
+
+    name: str
+    charm: str
+    resources: Optional[list[Resource]] = None
+    scale: int = 1
+    trust: bool = False
+    channel: Optional[str] = None
+
+
+class Relation(BaseModel):
+    """Pydantic model describing a charm relation."""
+
+    app_1_name: str
+    app_1_relation_name: str
+    app_2_name: str
+    app_2_relation_name: str
+
+
+class CharmBundle:
+    """Class describing a charm bundle."""
+
+    def __init__(
+        self,
+        name: str,
+        description: str,
+        applications: list[Application],
+        relations: Optional[list[Relation]],
+        bundle: str = "kubernetes",
+    ):
+        self.name = name
+        self.description = description
+        self.bundle = bundle
+        self.applications = applications
+        self.relations = relations
+
+    def render(self, output_file: str):
+        """Renders charm bundle using jinja2 templating.
+
+        Args:
+            output_file: Output file
+        """
```

Comment:
It might be interesting to also have a way to return the raw content instead of only writing to a file. Maybe having a `render` and `render_to_file` method?

Reply:
Done, there is not a `render` and a `render_to_file`.